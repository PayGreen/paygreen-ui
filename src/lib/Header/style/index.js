import styled from 'styled-components';

const HeaderBase = styled.header`
    
`;

export { HeaderBase };