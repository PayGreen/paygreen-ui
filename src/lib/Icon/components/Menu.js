import React from "react";

const SvgMenu = props => (
  <svg viewBox="0 0 143 143" {...props}>
    <path d="M114.8 45.6H26.2c-1.9 0-3.5-1.6-3.5-3.5s1.6-3.5 3.5-3.5h88.7c1.9 0 3.5 1.6 3.5 3.5s-1.6 3.5-3.6 3.5zM114.8 76.9H26.2c-1.9 0-3.5-1.6-3.5-3.5s1.6-3.5 3.5-3.5h88.7c1.9 0 3.5 1.6 3.5 3.5s-1.6 3.5-3.6 3.5zM114.8 107.9H26.2c-1.9 0-3.5-1.6-3.5-3.5s1.6-3.5 3.5-3.5h88.7c1.9 0 3.5 1.6 3.5 3.5s-1.6 3.5-3.6 3.5z" />
  </svg>
);

export default SvgMenu;
