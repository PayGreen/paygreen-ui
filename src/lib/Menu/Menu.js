import React from 'react';
import PropTypes from 'prop-types';
import { itemHtmlTagOptions, itemHtmlTagDefault } from '../../shared/constants';
import { MenuBase } from './style';

const Menu = props => {
    return (
        <MenuBase as={props.htmlTag} {...props}>
            {props.children}
        </MenuBase>
    );
};

Menu.propTypes = {
    htmlTag: PropTypes.oneOf(Object.values(itemHtmlTagOptions)),
};

Menu.defaultProps = {
    htmlTag: itemHtmlTagDefault,
};

export default Menu;
