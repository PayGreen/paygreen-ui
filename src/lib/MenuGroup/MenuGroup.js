import React from 'react';
import PropTypes from 'prop-types';
import { colorTypeOptions } from '../../shared/constants';
import Dot from '../Dot/Dot';
import MenuPrimary from '../MenuPrimary/MenuPrimary';
import MenuSecondary from '../MenuSecondary/MenuSecondary';
import MenuMobile from '../MenuMobile/MenuMobile';
import { MenuGroupBase } from './style';

const MenuGroup = ({ children, ...rest }) => {
    let hasOpenMenu = false;

    React.Children.map(children, child => {
        if (!child) {
            return null;
        }

        if ([MenuPrimary, MenuSecondary].includes(child.type)) {
            if (child.props.isOpen) {
                hasOpenMenu = true;
            }
        }
    });

    return (
        <MenuGroupBase hasOpenMenu={hasOpenMenu} {...rest}>
            {React.Children.map(children, (child, index) => {
                switch (child && child.type) {
                    case null:
                        return null;

                    case MenuMobile:
                        return React.cloneElement(child, {
                            isHidden: rest.isHidden && !hasOpenMenu,
                            hasTopStyle: rest.hasTopStyle && !hasOpenMenu,
                            key: index,
                        });

                    case Dot:
                        return React.cloneElement(child, {
                            colorType: rest.hasTopStyle
                                ? colorTypeOptions.reverse
                                : colorTypeOptions.original,
                            key: index,
                        });

                    default:
                        return child;
                }
            })}
        </MenuGroupBase>
    );
};

MenuGroup.propTypes = {
    isHidden: PropTypes.bool,
    hasTopStyle: PropTypes.bool,
};

MenuGroup.defaultProps = {
    isHidden: false,
    hasTopStyle: false,
};

export default MenuGroup;
