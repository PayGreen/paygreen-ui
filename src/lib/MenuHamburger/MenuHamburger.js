import React from 'react';
import PropTypes from 'prop-types';
import {
    arrowDirectionOptions,
    arrowDirectionDefault,
    colorPalletOptions,
    colorPalletDefault,
    colorThemeOptions,
    colorThemeDefault,
    formStatusOptions,
    formStatusDefault,
    greyOptions,
    greyDefault,
    spaceOptions,
    spaceDefault
} from '../../shared/constants';
import { MenuHamburgerBase } from './style';

const MenuHamburger = props => (
    <MenuHamburgerBase {...props}>
        <span></span>
        <span></span>
        <span></span>
    </MenuHamburgerBase>
);

MenuHamburger.propTypes = {
    arrowDirection: PropTypes.oneOf(Object.values(arrowDirectionOptions)),
    isOpen: PropTypes.bool,
    colorPallet: PropTypes.oneOf(Object.values(colorPalletOptions)),
    colorTheme: PropTypes.oneOf(Object.values(colorThemeOptions)),
    colorWab: PropTypes.oneOf(Object.values(greyOptions)),
    colorStatus: PropTypes.oneOf(Object.values(formStatusOptions)),
    marginTop: PropTypes.oneOf(Object.values(spaceOptions)),
    marginRight: PropTypes.oneOf(Object.values(spaceOptions)),
    marginLeft: PropTypes.oneOf(Object.values(spaceOptions)),
    marginBottom: PropTypes.oneOf(Object.values(spaceOptions)),
};

MenuHamburger.defaultProps = {
    arrowDirection: arrowDirectionDefault,
    isOpen: false,
    colorPallet: colorPalletDefault,
    colorTheme: colorThemeDefault,
    colorWab: greyDefault,
    colorStatus: formStatusDefault,
    marginTop: spaceDefault,
    marginRight: spaceDefault,
    marginLeft: spaceDefault,
    marginBottom: spaceDefault,
};

export default MenuHamburger;
