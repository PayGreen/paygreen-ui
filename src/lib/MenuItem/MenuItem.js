import React from 'react';
import PropTypes from 'prop-types';
import {
    colorThemeOptions,
    colorThemeDefault,
    alignOptions,
    alignDefault,
} from '../../shared/constants';
import { MenuItemBase } from './style';

const MenuItem = props => {
    return <MenuItemBase {...props}>{props.children}</MenuItemBase>;
};

MenuItem.propTypes = {
    colorTheme: PropTypes.oneOf(Object.values(colorThemeOptions)),
    isMain: PropTypes.bool,
    hasHoverBase: PropTypes.bool,
    hasHoverAlternative: PropTypes.bool,
    hasHoverEmphasis: PropTypes.bool,
    align: PropTypes.oneOf(Object.values(alignOptions)),
};

MenuItem.defaultProps = {
    colorTheme: colorThemeDefault,
    isMain: false,
    hasHoverBase: true,
    hasHoverAlternative: true,
    hasHoverEmphasis: false,
    align: alignDefault,
};

export default MenuItem;
