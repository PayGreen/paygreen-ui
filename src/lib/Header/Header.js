import React, { useState } from 'react';
import { useScrollPosition } from '@n8tb1t/use-scroll-position';
import { debounceTime } from '../../shared/constants';
import MenuGroup from '../MenuGroup/MenuGroup';
import MenuTertiary from '../MenuTertiary/MenuTertiary';
import { HeaderBase } from './style';

const Header = ({ children, ...rest }) => {
    const direction = {
        top: 'top',
        bottom: 'bottom',
    };
    const [scrollTop, setScrollTop] = useState(0);
    const [scrollDirection, setScrollDirection] = useState(direction.top);

    useScrollPosition(
        ({ prevPos, currPos }) => {
            setScrollTop(currPos.y);
            setScrollDirection(
                currPos.y > prevPos.y ? direction.bottom : direction.top,
            );
        },
        [],
        null,
        true,
        debounceTime,
    );

    let hasMenuTertiary = false;

    React.Children.map(children, child => {
        if (child && child.type === MenuTertiary) {
            hasMenuTertiary = true;
        }
    });

    const menuTertiaryStartDisplay = 110; // pixels to the top

    return (
        <HeaderBase {...rest}>
            {React.Children.map(children, child => {
                if (!child) {
                    return null;
                }

                if (child.type === MenuGroup) {
                    return React.cloneElement(child, {
                        isHidden:
                            hasMenuTertiary &&
                            scrollDirection === direction.bottom &&
                            scrollTop > menuTertiaryStartDisplay,
                        hasTopStyle: scrollTop <= 0,
                    });
                }

                if (child.type === MenuTertiary) {
                    return React.cloneElement(child, {
                        isHidden: scrollTop <= menuTertiaryStartDisplay,
                    });
                }

                return child;
            })}
        </HeaderBase>
    );
};

export default Header;
