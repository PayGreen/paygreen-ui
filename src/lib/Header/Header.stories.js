import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean } from '@storybook/addon-knobs';
import {
    folder,
    spaceOptions,
    alignOptions,
    colorPalletOptions,
    greyOptions,
    itemHtmlTagOptions,
    iconHtmlTagOptions,
    iconSizeOptions,
} from '../../shared/constants';
import { UserIcon, LockIcon } from '../Icon/Icon';
import Link from '../Link/Link';
import MenuItem from '../MenuItem/MenuItem';
import Menu from '../Menu/Menu';
import MenuPrimary from '../MenuPrimary/MenuPrimary';
import MenuSecondary from '../MenuSecondary/MenuSecondary';
import MenuTertiary from '../MenuTertiary/MenuTertiary';
import MenuClose from '../MenuClose/MenuClose';
import MenuHamburger from '../MenuHamburger/MenuHamburger';
import Logo from '../Logo/Logo';
import Dot from '../Dot/Dot';
import MenuGroup from '../MenuGroup/MenuGroup';
import Header from './Header';
import { logo } from '../Logo/sample/logo';

const mainLink = {
    hasHover: false,
    hasUnderline: false,
    colorPallet: colorPalletOptions.wab,
    hasUppercase: true,
};

const sampleLinks = [
    {
        link: 'lorem',
        label: 'Lorem ipsum dolor',
    },
    {
        link: 'sit-amet',
        label: 'Sit amet',
    },
    {
        link: 'consectetur',
        label: 'Consectetur adipiscing',
    },
    {
        link: 'ornare',
        label: 'Cras ornare laoreet',
    },
    {
        link: 'in-finibus',
        label: 'In finibus dolor',
    },
];

const isOpenMenuPrimaryLabel = 'Open MenuPrimary';

storiesOf(folder.nav + 'Header', module)
    .addDecorator(withKnobs)
    .add('Header', () => (
        <Header>
            <MenuGroup>
                <div className="main-nav">
                    <MenuHamburger
                        isOpen={boolean(isOpenMenuPrimaryLabel, false)}
                    />

                    <a href="#">
                        <Logo
                            hasBaseline={false}
                            hasHoverRight={true}
                            blockWidth={spaceOptions.sm}
                        >
                            {logo}
                        </Logo>
                    </a>

                    <UserIcon
                        htmlTag={iconHtmlTagOptions.button}
                        iconSize={iconSizeOptions.xl}
                    />
                </div>

                <MenuPrimary isOpen={boolean(isOpenMenuPrimaryLabel, false)}>
                    <Menu>
                        <a href="#">
                            <MenuItem isMain={true}>
                                <Link {...mainLink}>About us</Link>
                            </MenuItem>
                        </a>
                    </Menu>

                    <Menu>
                        <a href="#">
                            <MenuItem isMain={true}>
                                <Link {...mainLink}>Our products</Link>
                            </MenuItem>
                        </a>
                    </Menu>

                    <Menu>
                        <a href="#">
                            <MenuItem isMain={true}>
                                <Link {...mainLink}>Visit us</Link>
                            </MenuItem>
                        </a>
                    </Menu>
                </MenuPrimary>

                <Dot
                    className="hideOnSmallScreen"
                    colorPallet={colorPalletOptions.wab}
                    colorWab={greyOptions.grey10}
                    marginLeft={spaceOptions.lg}
                    marginRight={spaceOptions.lg}
                />

                <MenuSecondary
                    closeComponent={<MenuClose />}
                    isOpen={boolean('Open MenuSecondary', false)}
                >
                    <Menu htmlTag={itemHtmlTagOptions.div}>
                        <a href="#">
                            <MenuItem
                                hasHoverBase={false}
                                hasHoverEmphasis={true}
                                align={alignOptions.center}
                                isMain={true}
                            >
                                <LockIcon
                                    iconSize={iconSizeOptions.lg}
                                    colorPallet={colorPalletOptions.wab}
                                />

                                <Link
                                    hasHover={false}
                                    hasUnderline={false}
                                    hasUppercase={true}
                                    colorPallet={colorPalletOptions.wab}
                                >
                                    Login
                                </Link>
                            </MenuItem>
                        </a>
                    </Menu>
                </MenuSecondary>
            </MenuGroup>

            {boolean('Add MenuTertiary', true) ? (
                <MenuTertiary>
                    {sampleLinks.map((sample, index) => (
                        <a href={'#' + sample.link} key={index}>
                            {sample.label}
                        </a>
                    ))}
                </MenuTertiary>
            ) : null}
        </Header>
    ));
