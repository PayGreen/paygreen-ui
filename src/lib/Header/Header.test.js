import React from 'react';
import TestRenderer from 'react-test-renderer';
import { ThemeDefault } from '../../theme';
import MenuItem from '../MenuItem/MenuItem';
import Menu from '../Menu/Menu';
import MenuPrimary from '../MenuPrimary/MenuPrimary';
import MenuHamburger from '../MenuHamburger/MenuHamburger';
import MenuMobile from '../MenuMobile/MenuMobile';
import MenuGroup from '../MenuGroup/MenuGroup';
import Header from './Header';

it('renders without crashing', () => {
    const component = TestRenderer.create(
        <Header theme={ThemeDefault}>
            <MenuGroup theme={ThemeDefault}>
                <MenuMobile theme={ThemeDefault}>
                    <MenuHamburger theme={ThemeDefault} />
                </MenuMobile>

                <MenuPrimary theme={ThemeDefault}>
                    <Menu theme={ThemeDefault}>
                        <a href="#">
                            <MenuItem theme={ThemeDefault}>About us</MenuItem>
                        </a>
                    </Menu>
                </MenuPrimary>
            </MenuGroup>
        </Header>,
    );
    expect(component.toJSON()).toMatchSnapshot();
});
