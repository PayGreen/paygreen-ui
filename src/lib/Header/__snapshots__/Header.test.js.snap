// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<header
  className="sc-gZMcBi iSMSZB"
>
  <div
    className="main-nav"
  >
    <button
      className="sc-gzVnrw klyhST"
    >
      <span />
      <span />
      <span />
    </button>
  </div>
  <ul
    className="sc-bZQynM ikPvEl"
  >
    <li
      className="sc-EHOje dkqcmg"
    >
      <a
        href="#"
      >
        <div
          className="sc-bxivhb mGqfH"
        >
          <span
            className="sc-bdVaJa etiQqm"
          >
            About us
          </span>
        </div>
      </a>
    </li>
  </ul>
</header>
`;
